### PR TITLE
Fix invalid behavior of MoveToAttribute methods

### LINKED ISSUE #34443: XmlNodeReader MoveToAttribute bug, leaves XmlReader in inconsistent state
When using `XmlNodeReader.MoveToAttribute` returns false it leaves the XmlReader in an inconsistent state which can lead to follow-up exceptions.

Use-case: we are calling some deserialization API not owned by us which only accepts XmlReader. We already have an in-memory XmlDocument and need to call the deserialization method on a lot of nodes, so we want to use XmlNodeReader instead of doing an unnecessary format/parse roundtrip through an intermediate string for every node we have to deserialize.

In some cases the code calls `MoveToAttribute` to detect whether an optional attribute is present. This part works, but when it then subsequently calls `ReadStartElement` this leads to an exception. See attached example which reduces the call sequence to present the problem.

The error happens on both Desktop Framework and .NET Core.

```csharp
using System;
using System.IO;
using System.Linq.Expressions;
using System.Xml;

class Program
{
    static void Main(string[] args)
    {
        var doc = new XmlDocument();
        doc.LoadXml("<root><child attr1='value1'><other /></child></root>");

        Console.WriteLine("Testing XmlReader.Create");
        using (var rd = XmlReader.Create(new StringReader(doc.DocumentElement.FirstChild.OuterXml)))
            RunTest(rd);
        Console.WriteLine();

        Console.WriteLine("Testing XmlTextReader");
        using (var rd = new XmlTextReader(new StringReader(doc.DocumentElement.FirstChild.OuterXml)))
            RunTest(rd);
        Console.WriteLine();

        Console.WriteLine("Testing XmlNodeReader");
        using (var rd = new XmlNodeReader(doc.DocumentElement.FirstChild))
            RunTest(rd);
        Console.WriteLine();
    }

    static void RunTest(XmlReader rd)
    {
        Console.WriteLine($"  reader state: {rd.NodeType} {rd.Name}");
        Console.WriteLine($"IsStartElement(child) = {rd.IsStartElement("child")}");
        Console.WriteLine($"  reader state: {rd.NodeType} {rd.Name}");
        Console.WriteLine($"MoveToAttribute(attr1) = {rd.MoveToAttribute("attr1")}");
        Console.WriteLine($"  reader state: {rd.NodeType} {rd.Name}");
        Console.WriteLine($"MoveToAttribute(attr2) = {rd.MoveToAttribute("attr2")}");
        Console.WriteLine($"  reader state: {rd.NodeType} {rd.Name}");
        try
        {
            rd.ReadStartElement("child");
            Console.WriteLine($"ReadStartElement succeeded");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"ReadStartElement failed: {ex.Message}");
        }
    }
}
```

```text
Testing XmlReader.Create
  reader state: None
IsStartElement(child) = True
  reader state: Element child
MoveToAttribute(attr1) = True
  reader state: Attribute attr1
MoveToAttribute(attr2) = False
  reader state: Attribute attr1
ReadStartElement succeeded

Testing XmlTextReader
  reader state: None
IsStartElement(child) = True
  reader state: Element child
MoveToAttribute(attr1) = True
  reader state: Attribute attr1
MoveToAttribute(attr2) = False
  reader state: Attribute attr1
ReadStartElement succeeded

Testing XmlNodeReader
  reader state: None
IsStartElement(child) = True
  reader state: Element child
MoveToAttribute(attr1) = True
  reader state: Attribute attr1
MoveToAttribute(attr2) = False
  reader state: Element attr1
ReadStartElement failed: Element 'child' was not found.
```

### PR DESCRIPTION
**Pull request by Maximys**
This PR fix [Issue-34443](https://github.com/dotnet/runtime/issues/34443).

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlNodeReader.cs [hunk 1]
```diff
@@ -1395,6 +1386,7 @@ public override bool MoveToAttribute(string name, string? namespaceURI)
                 return true;
             }
             _readerNav.RollBackMove(ref _curDepth);
+            _nodeType = savedNodeType;
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlNodeReader.cs:1389]: do `_curDepth` or `_bInReadBinary` need any adjustments? (I've done quick tests with checking `XmlReader.Depth` and that looks consistent across readers so possibly the answer is no)

**Review by krwq — APPROVED** [review]

**Comment by krwq** [comment]
@Maximys I gave this a second look and per https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlreader.movetoattribute?view=net-6.0#System_Xml_XmlReader_MoveToAttribute_System_String_ the single string MoveToAttribute can take a qualified name while (localName, ns) always explicitly takes two. I.e. see implementation https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextReaderImpl.cs#L1043 here - can you add tests to ensure this scenario works?

**Comment by krwq** [comment]
@Maximys have you perhaps had a chance to test out what I suggested above?

**Comment by Maximys** [comment]
@krwq , am I correctly understand you, that your want to check is next XML will read fine or not:
`"<root><child someNamespace:attr1='value1'><other /></child></root>"`

**Comment by danmoseley** [comment]
@krwq  question for you above

**Comment by krwq** [comment]
@Maximys had a chance to look in the scenario I mentioned above?

**Comment by danmoseley** [comment]
I'm going to close this due to not hearing back. Feel free to reopen if you wish to continue.

**Pull request closed without merging**